### ISSUE TITLE
Support various grab input styles

### ISSUE THREAD

**Issue by dantman** [thestonefox/VRTK#1393]
I understand VRTK plans to eventually move away from the fixed mapping of controller buttons to actions that can be customized. That should help solve the currently (from what I can see) limited variety of grab input options.

When designing what new actions there are to be used by the grab scripts and how they are fired this variety of possible grab input options should be considered. Namely it should be considered that different controllers have different ideal techniques for how to handling grab (so customization by that is desirable) and it may be desirable to use multiple grab methods in the same scene to interact with different classes of object.

## Oculus Touch

* Generic grabbing:
  * Hand trigger press = Grip start
  * Hand trigger release = Grip release
* *Others unknown*

## Vive wands

Large objects:
* Grip toggle:
  * Grip button press = Grip start
  * Grip button press when grabbing = Grip release
* Grip hold:
  * Grip press = Grip start
  * Grip release = Grip end
* Trigger hold:
  * Trigger press = Grip start
  * Trigger release = Grip end
* Drop button:
  * Trigger button press = Grip start
  * Grip button press = Grip end

These various styles of grabbing can sometimes be combined in games:
* "Grip hold" for short-term objects and "Grip toggle" for longer-term tools
* "Trigger hold" for short-term objects and "Drop button" for longer-term tools

## Knuckles

Knuckles can be used much like Touch or the Vive wands but also has its own unique gestures.

Judging by the Knuckles interaction demo the controller seems well suited to "pinching" objects, which seems to be a gesture combining the trigger's sensor and either the trackpad or face button's capacitive sensor. Additionally instead of a grip button or hand trigger axis an average of the curl of the 3 fingers on the grip is used.

* Pinching
  * Pinching started = Grip start
  * Pinching stopped = Grip ended
* Hand curl
  * Average grip finger curl past threshold = Grip start
  * Average grip finger curl left threshold = Grip end

Like with the wands it may be desirable to use these together in the same game:
* "Hand curl" for large tools with a grip (guns, swords, and other tools), "Pinching" for small objects (picking up objects <5cm in diameter).

**Comment by thestonefox** [comment]
The knuckles aren't suited for pinching from my time with using them.

A pinch would be no different to putting your finger on the trigger and your thumb on the touchpad, which would be the same motion as you do on a vive wand.

Other than that the above actions all seem the same albeit using a different button for grabbing (except in the case where the trigger picks up and the grip drops, but you'd achieve that at the moment by just swapping the button on grab and on release)

**Comment by dantman** [comment]
> The knuckles aren't suited for pinching from my time with using them.

Two questions:
* Have you tried using the [Knuckles demo scene](https://steamcommunity.com/sharedfiles/filedetails/?id=953648265) which seems to do primary grabbing by pinching
* Have you tried doing pinching not by the trackpad but by a face button (inner face button?). I keep hearing from developers with the Knuckles that it feels like a natural pinching gesture ([one of the responses I can find](https://www.reddit.com/r/Vive/comments/6l1kos/nextgen_steamvr_knuckles_megathread_qa/djqfb3b/)).

We're describing the same thing by finger curl sensing and open/closed hand based grip. Axis 2.x has an average of the curl of the fingers and when that average goes past a threshold you'd get "grip start" and when that average drops back below the threshold you get "grip release".

> Other than that the above actions all seem the same albeit using a different button for grabbing (except in the case where the trigger picks up and the grip drops, but you'd achieve that at the moment by just swapping the button on grab and on release)

Ok, so there is a hack to make that type of interaction possible in the current system.

I could probably summarize what I'm hoping for in a list rather than trying to describe it with a list of control methods:

* @thestonefox You mentioned you plan to switch from button mapping to some form of action events for controller inputs based of of Unity's work on something similar (either using their system when it comes out or making a simple version of it in VRTK if it's easy to do and you have time).
* I'm hoping to influence how grabbing controls are handled in that new system, so that it takes a few different factors into account.
  1. On the Touch controllers holding in the hand trigger works fine for holding on to things, but on the Vive wands this cramps peoples hands so a grip toggle is better. *So I hope it's taken into account that different controllers have different preferred inputs.*
  2. Sometimes the desired grip input involves more than one button. *Like the trigger to grab and grip to drop.*
  3. One grip method is not always enough. *Like when you want to use the grip to drop method on your tools and another method (grip hold or trigger hold) for grabbing other objects.*
* Right now interactable objects are in charge of whether you hold the button to grab or it toggles. Which doesn't feel right. On Oculus Touch and Knuckles you may want to hold the grip for all objects instead of toggling them while you use grip toggle in the same scene if a Vive wand is used.
* Here's what I'm hoping for in the new system based off these factors:
  * The action system should have separate "grip/grab start" and "grip/grab released" actions/events. And the script that emits those can map them to completely different things (they can be mapped to the same button press so you can do a grip toggle, or release can be mapped to the release of a button, or start and release can be mapped to two different buttons). And Interaction grab/objects won't care about how this grab input works.
  * The fact that each controller type has a different grip method that is ideal should be taken into account, especially whether it's ok to hold the grip button or not. So instead of being forced to write a script to change the options in VRTK_ControllerEvents the new system should have a built-in way to use different options when certain controllers are in use. (I believe you already have something similar for per-controller offsets).
  * The system should have at least two classes of grab input, to take into account the fact that you may want to use different input for grabbing different objects (one for tools, one for throwable objects). You'd use this instead of the InteractableObject deciding whether you have to hold the button or not. This way you can still do grip toggle on the Vive (Grab class A = grip toggle; Grab class B = grip hold), but on Oculus Touch and Knuckles you can map hand trigger/grip hold to both classes, and it's possible to also set the Vive wands to grab combinations other than toggle vs. hold (Grab class A = trigger press to hold, grab to drop; Grab class B = trigger hold to grab).

**Comment by thestonefox** [comment]
@dantman 

Yeah I've tried the interaction demo, it's a good example of how not to use the knuckles controllers. They just don't have the fidelity required to pull that off. Whoever built that demo probably had ideal hand size for the controllers and was very careful when picking things up.

The knuckles suck if your hands are too small or too large (e.g. I can never get a good hand curl because it always thinks my pinky and ring finger are the same finger).

Plus that demo has a rigidbody sphere collider being updated every frame which is crazy inefficient.

In reality, the knuckles are a nice idea, but from my time using them they're not ready yet or they're just not going to live up to the hype.

Which is the same as the pinching, again I think this is another hype train by some devs who maybe pinching feels good for them, but for many other people just doesn't feel natural. Having to press two buttons to mimic a pinch isn't any more natural than having to press one button to grab. The position of the trigger and buttons means your fingers aren't even doing a pinch motion but some weird finger curl. You could force yourself into doing a pinch motion but it's not natural.

Now on to the button action mapping.

You can already achieve everything you're asking for now, albeit you'd have to write custom code to achieve swapping of buttons based on situation and controller type.

The new action system isn't going to have anything baked in that would deal with grabbing specifically.

The point of the new action system would be that a pre-written action would throw an event when the action occured and this event would be used in place of the existing button press.

So with grabbing, you still would have the same set up as you do now, but rather than specific "grip" to be the grab button, you would specify an action listener as the grab action.

Now this action listener could have all sorts of customisation to it.

e.g. it could have a button press event just like now, so grip button raises the action to grab, and releasing the grip button would release the object (unless the toggle grab was turned on).

But you could make it more complicated by saying grip and trigger need to be pressed, or something like grip or trigger can be pressed.

You would also be able to listen for other events within the action listener, e.g. when you touch an interactable object, get the object script you're touching, determine which controller you're using and specify the button you want for grabbing.

All of these action listeners will form their own script that you would pick and choose the most appropriate one (similar to how the object control actions work now)

so, in short, the action listeners won't specifically have anything to do with gripping in them, but you will certainly be able to write something generic and reusable to do that.

**Comment by treefortress** [comment]
Running into a semi-related issue. At the core of it, I want to be able to initiate a grab, anytime the grab item is held down, just to make it easier to pick things up with a quick swiping motion, or to easily enable telekinetic powers (draw the object towards the hand while grab is held, hand automatically grab when objects comes in range). 

Another use case is climbing, if you're falling down a wall, and there's various hand holds flying by you, it's much easier to just stick out your hands with the Grab button held down, and hope to catch a ledge, than it is to try and time your ButtonDown presses just right.

So that's one issue, **could there be a core update to allow for "Grab while holding"?**

Currently I can override VRTK_InteractGrab, and basically get what I need, but that does bring me to another issue, and I'm not sure this is worth a ticket.... **Would it be possible to completely replace all private member variables with protected?** It's very frustrating when you can't access something because it's private. Case in point, to get my grab-while-held behavior I want to override VRTK_InteractGrab.OnEnable, and hook onto the controllerEvents.AliasGrabOn, so I can make repeated calls to AttemptGrab, but I can't access controllerEvents because it's private. It's nice that the functions are marked as virtual, so we can override them, but with all the member vars being private, it makes it hard to do anything.

**Comment by thestonefox** [comment]
@treefortress you can already do this with the grab precongnition setting on the interact grab script.

It will attempt to grab if the grab button is pressed x number of seconds before the touch occurs

The climbing example scene already uses this to make it easier to climb

**Comment by treefortress** [comment]
Ya I realized this after posting, but still a bit weird cause I'd have to put an extremely high value in there to get the behavior I want. Which would work, just seems a bit hacky. Overriding it was easy enough and seems to work fairly well. 

Thoughts on defaulting to protected over private across the lib? Would just make it easier to extend...

**Comment by thestonefox** [comment]
@treefortress what is private for you? There should be no private things any more

**Comment by treefortress** [comment]
Apologies, I'm working off a version that is a couple of mths old, looks like you're way ahead of me :)
Cheers,

**Comment by thestonefox** [comment]
Linked to https://github.com/thestonefox/VRTK/issues/1523
